guis/panel/heater/load.py

```diff
@@ -0,0 +1,91 @@
+import guis.panel.heater.load_heat_csv_into_db as load, os, sys, pandas as pd, shutil, time
+from pathlib import Path
+
+def copy_csv(filename):
+    rootdir = "data/Panel data/external_gui_data/heat_control_data/"
+    copy_path = "guis/panel/heater/problem_files/"
+    original = rootdir + filename
+    target = copy_path + filename
+    shutil.copyfile(original, target)
+
+if __name__ == "__main__":
+    
+    # Get list of all files
+    rootdir = "data/Panel data/external_gui_data/heat_control_data/"
+    dirs = os.listdir(rootdir)
+    date_format = "%Y-%m-%d"
+    
+    # Make dict {panel : (timestamp, process, file)}
+    heat_files_info = {}
+    bad_files = []
+    for f in dirs:
+        f = f.replace(".csv", "")
+        parsed_filename = f.split("_")
+    
+        # Add messed up files to bad list and move on
+        if len(parsed_filename) not in [2,3]:
+            bad_files.append(f)
+            continue
+    
+        # Parse info from filename
+        panel_number = parsed_filename[0]
+        date = parsed_filename[1]
+        process = None if len(parsed_filename) == 2 else parsed_filename[2]
+    
+        # Validate parsed info
+        # Send bad files to a bad list, add good entries to the dictionary
+        try:
+            panel_number = int(panel_number[2:]) # remove "MN", convert to int
+            assert panel_number < 200 # require panel number < 200
+            date = time.strptime(date, date_format)
+            
+        except: # catch invalid panel number or date format
+            bad_files.append(f)
+        else:
+            if panel_number not in heat_files_info:
+                heat_files_info[panel_number] = []
+            heat_files_info[panel_number].append((date, process, f))
+            
+            
+    
+
```

Comment:
I know I drafted this block, but I don't actually like what I started.
1. As is, we're looping over the _key_ of a dictionary here. The key of our dictionary is the panel number -- so for starters, `key` should be renamed `panel_number`. I like to be specific whenever possible in loops.
2. `for i in heat_files_info[key]: bad_files.append(i[2])` this is hard to read. We can do a little bit better by looping over the _key_ and _value_ of our dict at once:
```
for panel_number, files_info in heat_files_info.items():
    if len(files_info) != 3:
        to_remove.append(panel_number)
        for file_info in files_info:
            filename = file_info[2] 
            bad_files.append(filename)
```
That's a little better.

3. Finally, I think I already expressed my fear of removing elements from lists. I think what's being done here
```
for i in to_remove:
    del heat_files_info[i]
```
 is actually gonna give the wrong answer! `del heat_files_info[i]` deletes the ith _index_ of `heat_files_info`. BUT we filled the `to_remove` list with panel numbers! So we'll try to remove the 138th index of `heat_files_info` when we meant to remove the entries for the panel number MN138. I'm surprised this did not throw a `KeyError` (i.e. because there aren't 138 entries in the `to_remove` list). I haven't actually run this code, so maybe I'm misinterpreting something.

`remove` is the one that removes an element from a list by _value_ instead of by index.

**Anyway here's my final draft, including some other small improvements.**

```
to_remove = [panel_number for panel_number, files_info in heat_files_info.items() if len(files_info) != 3]
for panel_number in to_remove:
    for file_info in heat_files_info[panel_number]:
        filename = file_info[2] 
        bad_files.append(filename)
    heat_files_info.remove(panel_number)
```